src/governance.cpp

```diff
@@ -902,71 +904,53 @@ bool CGovernanceManager::MasternodeRateCheck(const CGovernanceObject& govobj, up
         return false;
     }
 
-    bool fAdditionalRelay = false;
     if(nTimestamp > nNow + MAX_TIME_FUTURE_DEVIATION) {
         LogPrintf("CGovernanceManager::MasternodeRateCheck -- object %s rejected due to too new (future) timestamp, masternode vin = %s, timestamp = %d, current time = %d\n",
                  strHash, vin.prevout.ToStringShort(), nTimestamp, nNow);
         return false;
-    } else if (nTimestamp > nNow + MAX_TIME_FUTURE_DEVIATION - RELIABLE_PROPAGATION_TIME) {
-        // schedule additional relay for the object
-        fAdditionalRelay = true;
+    }
+
+    txout_m_it it  = mapLastMasternodeObject.find(vin.prevout);
+    if(it == mapLastMasternodeObject.end())
+        return true;
+
+    if(it->second.fStatusOK && !fForce) {
+        fRateCheckBypassed = true;
+        return true;
     }
 
     double dMaxRate = 1.1 / nSuperblockCycleSeconds;
     double dRate = 0.0;
     CRateCheckBuffer buffer;
-    CRateCheckBuffer* pBuffer = NULL;
     switch(nObjectType) {
     case GOVERNANCE_OBJECT_TRIGGER:
         // Allow 1 trigger per mn per cycle, with a small fudge factor
-        pBuffer = &it->second.triggerBuffer;
+        buffer = it->second.triggerBuffer;
```

Comment:
A local variable is used here to check whether the rate limit exceeded if the object is accepted. I moved the buffer updating outside of `MasternodeRateCheck` to a separate method `MasternodeRateUpdate`.